### DISCUSSION TITLE
Enhancing Home Assistant Backup(s)

### DISCUSSION THREAD

**Discussion by ludeeus** [home-assistant/architecture#1134] [General] [answered]
## Context

Currently, we have 2 very different ways of handling backups:
- The core backup integration
- The supervisor (hassio)

When we add cloud backups, we want to do so in a way that other integrations can reuse to provide more flexibility. We all know that more is better for backups.

## Step 1: Backup sync agents

Backup sync agents build upon the existing [backup platform](https://developers.home-assistant.io/docs/core/platform/backup) within the core. This extension introduces the ability for integrations to register sync agents that will be triggered upon backup completion. 

Example implementation in the demo integration: https://github.com/home-assistant/core/commit/bb1499e5da373c762b26d48a0e66fbc3f4265cf1

Each integration wanting to support this feature will implement the `async_get_backup_agents` function in their backup platform. This function will return a list of instances that inherit from the `BackupSyncAgent` base class. These agents must implement the following methods:
- `async_download_backup` will download a previously synced backup
- `async_upload_backup` will upload the backup (the path to the backup file would be inside the `Backup` object)
- `async_list_backups` will list all synced backups if any.

The backup integration will then have new handlers to:
- List active agents
- List stored backups for all agents
- Trigger a download of a backup for a specified agent.
- Send updates over the websocket when the state/stages of agent changes

On a system with a supervisor, we need to extend the environment:
- The core container needs direct access to the generated archive
- The supervisor needs to signal core when the archive is created

## Step 2: Proxy hassio calls

While the placement in the UI is somewhat similar, they are completely different (one shows the native backup integration, and the other an embedded panel hosted by the supervisor).

To make these experiences similar we will implement all actions that the hassio integration currently set up for handling backups in the backup integration.
But if an hassio exists, this integration will just proxy the action over to the hassio integration so that it will still do it like before.

- The supervisor will still be responsible for creating the backup, this is only to make the user experience and the triggering of the backing up as close as possible.
- This will make it possible to tell the user the same steps regardless on how the user is running Home Assistant.
- This will also make it possible for the backup integration to invoke its backup sync agents once the backup is completed.
For this to work as expected, the core container would need access to the newly created backup, so we do not need to stream this between the containers.
- We would also need to ensure that the supervisor's API does not return early or that we receive an event when the backup is complete.

## Step 3: Core restore

This brings us over to the last step, to fully have a consistent backup platform we need to be able to handle restore on core/container installations.

There are 2 forms of restoring that will apply to these 2:
- Restore Home Assistant Core during onboarding
- Restore Home Assistant Core after onboarding

Because a user have more freedom on how the environment is setup for these two we need to have some rules, and MVP lets tackle the first one only (restore during onboarding).
- Restore as a rollback would not be possible
  - As we can not change Home Assistant's runtime, we can not roll back a version with a backup; we can only move forward.
- Restore can only be done if:
  - Version is the same or newer
  - Onboarding has not been started (or completed)
  - There are no additional directories or files in the configuration directory
  - If a user maps in a file/directory we need to abort.

### So how will this work?

One option is to take inspiration from the safe mode, which writes a special file that is read during boot.
We can do the same here, add a `.ha_restore` file with the path to the backup archive and an optional password if the archive is encrypted.
If this file exists, we will not proceed with starting core but instead wipe the configuration directory (if the conditions are met) before extracting the contents of the archive to the same location.

**Comment by balloob** [comment]
This looks good.

Some feedback:

- `async_get_backup_agents` allows multiple. This is good as it allows multiple config entries to provide an agent
- Step 2 includes "But if an hassio exists, this integration will just proxy the action over to the hassio integration so that it will still do it like before.". Just a small implementation detail: instead of proxy, I would have 2 different agent classes. We insert the supervisor-powered one when available.
- "Restore as a rollback would not be possible". Why is this? Generally your backup is from an older version of Home Assistant and this would make it unusable?
- `.ha_restore` I like this.

**Reply by frenck** [reply]
Some other feedback:
- Backup progress/state; if a users goes to the backup page, we should be able to show it is backing up / syncing atm. The current model doesn't seem to reflect such possibility.
- Should the meta data be extended with information on the contents of the backup?

**Reply by ludeeus** [reply]
> We insert the supervisor-powered one when available.

I agree 👍 
> "Restore as a rollback would not be possible". Why is this?

We are not currently capable of doing so. In a container environment, you need access to the runtime socket for that.
If that is desired, I think a dedicated discussion around it would be needed, as that will change the current ADR (013).
For now, I would like to keep the focus here on providing what is required to get you back up and running if you have to start from scratch with a new install.

> Backup progress/state; 

The backing-up state is already implemented in both places. I'm not sure if they are actively being displayed, but the state is there.
As for syncing, yee that is my bad I left that out of the proposal, but yes it would be there on a global level as a start and we can later see if it is needed on a pr. agent later.

>Should the meta data be extended with information on the contents of the backup?

What metadata?
The async_upload_backup method would be called with an object representing the backup contents.

**Comment by frenck** [comment, marked as answer]
Other than the above comment, the proposal has been approved, and you can start working on the implementation while we resolve the last bit of the discussion 👍 

../Frenck
